helix-term/src/ui/context.rs

```diff
@@ -0,0 +1,510 @@
+        .take_while(|(i, _)| {
+            *i + Into::<usize>::into(config.sticky_context.indicator) != context.visual_row
+        }) // also only nodes that don't overlap with the visual cursor position
+        .map(|(i, node)| {
+            let mut new_node = node.clone();
+            new_node.visual_line = i as u16;
+            new_node
+        })
+        .collect();
+
+    if config.sticky_context.indicator {
+        res = add_indicator(&context.viewport, view, res);
+    }
+
+    Some(res)
+}
+
+fn build_cached_nodes(
+    nodes: &Option<Vec<StickyNode>>,
+    view: &View,
+    context: &mut StickyNodeContext,
+) -> Option<Vec<StickyNode>> {
+    if let Some(nodes) = nodes {
+        if nodes.iter().any(|node| view.id != node.view_id) {
+            return None;
+        }
+
+        // nothing has changed, so the cached result can be returned
+        if nodes.iter().any(|node| view.offset.anchor == node.anchor) {
+            return Some(nodes.iter().take(context.visual_row).cloned().collect());
+        }
+
+        // We're not on the same position as before, so instead we can reuse all the previous nodes.
+        let mut cached_nodes = nodes.to_vec();
+
+        // While the cached nodes are outside our search-range, pop them, too
+        cached_nodes.retain(|node| node.byte_range.start > context.topmost_byte);
+
+        return Some(cached_nodes);
+    }
+
+    None
+}
+
+fn get_context_paired_range(
+    query_match: &QueryMatch,
+    start_index: u32,
+    end_index: u32,
+    top_first_byte: usize,
+    last_scan_byte: usize,
+) -> Option<std::ops::Range<usize>> {
+    // get all the captured @context.params nodes
+    let end_nodes = once_cell::unsync::Lazy::new(|| {
+        query_match
+            .nodes_for_capture_index(end_index)
+            .collect::<Vec<_>>()
+    });
+
+    query_match
+        .nodes_for_capture_index(start_index)
+        .find_map(|context| {
+            let ctx_start_range = context.byte_range();
+
+            // filter all matches that are out of scope, based on follows-cursor
+            let start_range_contains_bytes = ctx_start_range.contains(&top_first_byte)
+                && ctx_start_range.contains(&last_scan_byte);
+            if !start_range_contains_bytes {
+                return None;
+            }
+
+            let ctx_start_row = context.start_position().row;
+            let ctx_start_byte = ctx_start_range.start;
+
+            end_nodes.iter().find_map(|it| {
+                let end = it.end_byte();
+                // check whether or not @context.params nodes are on different lines
+                (ctx_start_row != it.end_position().row && ctx_start_range.contains(&end))
+                    .then_some(ctx_start_byte..end.saturating_sub(1))
+            })
+        })
+}
+
+/// Tests whether or not a given node is in a specific tree-sitter range
+fn node_in_range(
+    node: Node,
+    anchor: usize,
+    node_byte_range: &Option<std::ops::Range<usize>>,
```

Comment:
```suggestion
    node_byte_range: Option<&std::ops::Range<usize>>,
```
Inspired by https://github.com/helix-editor/helix/pull/11091